lib/TPP/ConvertCheckToLoops.cpp

```diff
@@ -0,0 +1,133 @@
+//===- ConvertCheckToLoops.cpp ------------------------------------*- C++
+//-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "TPP/Dialect/Check/CheckOps.h"
+#include "TPP/Passes.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/ControlFlow/IR/ControlFlowOps.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Math/IR/Math.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace mlir;
+using namespace mlir::check;
+using namespace mlir::cf;
+using namespace mlir::arith;
+using namespace mlir::math;
+
+#define GEN_PASS_CLASSES
+#include "TPP/Passes.h.inc"
+
+namespace {
+
+// Convert 
+// check.expect_almost_equals(%t1:tensor<2x2xf32>, %t2:tensor<2x2xf32>,
+// %threshold:f32) 
+// to the pattern: 
+// scf.some_loop(%i, %j)
+// %0 = load from %t1 
+// %1 = load from %t2 
+// %diff = math.subf %0, %1 
+// %abs = arith.absf %diff 
+// %compare = arith.cmpf le, %abs, %threshold 
+// cf.assert %compare, "Result mismatch"
+struct ConvertAlmostEqualsOp
+    : public OpRewritePattern<check::ExpectAlmostEqOp> {
+  using OpRewritePattern<check::ExpectAlmostEqOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(check::ExpectAlmostEqOp almostEqOp,
+                                PatternRewriter &rewriter) const override {
+    Location loc = almostEqOp.getLoc();
+    SmallVector<Value> ubs;
+    if (!almostEqOp.getLhs().getType().isa<MemRefType>()) {
+      return failure();
+    }
+    size_t rank = almostEqOp.getLhs().getType().cast<MemRefType>().getRank();
+    for (size_t idx = 0; idx < rank; idx++) {
+      // Works for dynamic shapes
+      // If the index is a constant, the dim value is folded into a constant
+      Value dimIndex = rewriter.create<arith::ConstantIndexOp>(loc, idx);
+      Value dim = rewriter.create<memref::DimOp>(loc, almostEqOp.getOperand(0),
```

Comment:
Can this get folded? I was typically using `linalg::createFoldedDimOp`

Reply:
https://github.com/plaidml/tpp-sandbox/blob/main/lib/TPP/TransformUtils.cpp#L147

Reply:
This automatically folds if the dimIndex is a constant, from what I've noticed.